src/sentry/lang/native/symbolicator.py

```diff
@@ -192,6 +192,69 @@ def process_payload(self, stacktraces, modules, signal=None):
         )
 
 
+def redact_internal_sources(response):
+    """Redacts information about internal sources from a response.
+
+    Symbolicator responses can contain a section about DIF object file candidates where were
+    attempted to be downloaded from the sources.  This includes a full URI of where the
+    download was attempted from.  For internal sources we want to redact this in order to
+    not leak any internal details.
+
+    Note that this modifies the argument passed in, thus redacting in-place.  It still
+    returns the modified response.
+    """
+    for module in response.get("modules", []):
+        redact_internal_sources_from_module(module)
+    return response
+
+
+def redact_internal_sources_from_module(module):
+    """Redacts information about internal sources from a single module.
+
+    This in-place redacts candidates from only a single module of the symbolicator response.
+
+    The strategy here is for each internal source to replace the location with the DebugID.
+    Furthermore if there are any "notfound" entries collapse them into a single entry and
+    only show this entry if there are no entries with another status.
+    """
+    sources_notfound = set()
+    sources_other = set()
+    new_candidates = []
+
+    debug_id = module.get("debug_id", "<missing-debug-id>")
+    for candidate in module.get("candidates", []):
+        source_id = candidate["source"]
+        if is_internal_source_id(source_id):
+            candidate["location"] = "debugid://{}".format(debug_id)
+            try:
+                status = candidate.get("download", {})["status"]
+            except KeyError:
+                new_candidates.append(candidate)
+            else:
+                if status == "notfound":
+                    if source_id in sources_notfound:
+                        continue
+                    else:
+                        sources_notfound.add(source_id)
+                else:
+                    sources_other.add(source_id)
+                new_candidates.append(candidate)
+        else:
+            new_candidates.append(candidate)
+
+    for i, candidate in enumerate(new_candidates.copy()):
+        source_id = candidate["source"]
+        try:
+            status = candidate.get("download", {})["status"]
+        except KeyError:
+            continue
+        if status == "notfound" and source_id in sources_other:
+            del new_candidates[i]
```

Comment:
This is modifying `new_candidates` by the _original_ index, which has changed after the first removal. Can you rewrite this using `filter`?